Block: Implement block insertion functions

ockham opened the pull request:
See https://core.trac.wordpress.org/ticket/59313 and #5158.

Trac ticket: https://core.trac.wordpress.org/ticket/59385

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ockham commented on tests/phpunit/tests/blocks/blockHooks.php:
```diff
@@ -97,4 +100,82 @@ public function test_get_hooked_blocks_matches_found() {
+	}
+
+	/**
+	 * @ticket XXXXXX
+	 */
+	public function test_inject_hooked_block_at_last_child_position() {
+		$original_markup = <<<HTML
+<!-- wp:tests/group-last-child {"layout":{"type":"constrained"}} -->
+	<div class="wp-block-group">
+		<!-- wp:paragraph -->
+			<p>Foo</p>
+		<!-- /wp:paragraph -->
+	</div>
+<!-- /wp:tests/group-last-child -->
+HTML;
+
+		$inserted_block = array(
+			'blockName'    => 'tests/hooked-block',
+			'attrs'        => array(),
+			'innerBlocks'  => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+		);
+
+		$blocks = parse_blocks( $original_markup );
+		$actual = insert_hooked_block( $inserted_block, 'last_child', $blocks[0] );
+
+		$expected_markup = <<<HTML
+<!-- wp:tests/group-last-child {"layout":{"type":"constrained"}} -->
+	<div class="wp-block-group">
+		<!-- wp:paragraph -->
+			<p>Foo</p>
+		<!-- /wp:paragraph --><!-- wp:tests/hooked-block /-->
+	</div>
+<!-- /wp:tests/group-last-child -->
+HTML;
+
+		$expected = parse_blocks( $expected_markup )[0];
+		$this->assertSame( $expected, $actual );
+	}
```
These tests might actually be data provider material: We can harmonize them so they all use the same `$original_markup` (i.e. `<!-- wp:tests/group -->` instead of `<!-- wp:tests/group-first-child -->` or `<!-- wp:tests/group-last-child -->`, respectively). The data provider then provides the position, and the corresponding expected resulting markup.

ockham replied:
Done in e8f5342.

ockham commented:
@gziolo I've decided to only implement `insert_block_hook` and `inser_block_hooks` here (plus unit test coverage). We'll only wire them to block templates and patterns to actually insert blocks in a follow-up 😄 

Edit: I've updated the [tracking ticket](https://core.trac.wordpress.org/ticket/59313) accordingly.

ockham commented on tests/phpunit/tests/blocks/blockHooks.php:
```diff
@@ -97,4 +100,144 @@ public function test_get_hooked_blocks_matches_found() {
+		<!-- /wp:paragraph -->
+	</div>
+<!-- /wp:tests/group -->
+HTML
```
PHP 7.0 didn't like the trailing comma 😕

gziolo replied:
Yes, that is a bit annoying. Usually, you need to put the string in its variable to use Heredoc:

> Warning
> Prior to PHP 7.3.0, it is very important to note that the line with the closing identifier must contain no other characters, except a semicolon (;). That means especially that the identifier may not be indented, and there may not be any spaces or tabs before or after the semicolon. It's also important to realize that the first character before the closing identifier must be a newline as defined by the local operating system. This is \n on UNIX systems, including macOS. The closing delimiter must also be followed by a newline.

So it's only the semicolon allowed which won't work here 😞

ockham commented:
@gziolo There's another fix we might have to backport to GB: e9931e7.

ockham commented:
@gziolo This needs test coverage for `insert_hooked_blocks` (plural) (see https://github.com/WordPress/wordpress-develop/pull/5241#discussion_r1328811091); otherwise, it might be about ready 🤞

ockham commented:
Will need https://github.com/WordPress/wordpress-develop/pull/5242 to eventually pass the parent arg to `insert_hooked_blocks`.

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,108 @@ function get_hooked_blocks( $name ) {
 	return $hooked_blocks;
 }
 
+function insert_hooked_blocks( $block ) {
+	$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+	foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+		$hooked_block = array(
+			'blockName'    => $hooked_block_type,
+			'attrs'        => array(),
+			'innerHTML'    => '',
+			'innerContent' => array(),
+			'innerBlocks'  => array(),
+		);
+		// Need to pass full current block, (possibly its parent), relative position, and hooked_block.
+		$block = insert_hooked_block( $hooked_block, $relative_position, $block );
+	}
+	return $block;
+}
```
This will need https://github.com/WordPress/wordpress-develop/pull/5242 to work with `serialize_blocks`.

```suggestion
function insert_hooked_blocks( $block, $parent ) {
	$hooked_blocks = get_hooked_blocks( $block['blockName'] );
	foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
		$hooked_block = array(
			'blockName'    => $hooked_block_type,
			'attrs'        => array(),
			'innerHTML'    => '',
			'innerContent' => array(),
			'innerBlocks'  => array(),
		);

		$block = insert_hooked_block( $parent, $hooked_block, $relative_position, $block );
	}
	return $block;
}
```

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,108 @@ function get_hooked_blocks( $name ) {
+ * designated parent block by inserting into the parent's `innerBlocks` array,
+ * and by updating the parent's `innerContent` array accordingly.
+ *
+ * @since 6.4.0
```
It's possible to mark it as private which will exclude it from the documentation:
```suggestion
 * @since 6.4.0
 * @access private
```

ockham replied:
I was actually considering keeping it public 🤔 It's not like I expect people to use it a lot, but it's going to be part of a few call stacks (hooked block insertion for both templates and patterns) so I thought it makes sense to have it documented.

gziolo replied:
Works for me. In that case the following line should go away:

>  It is not meant for public use.

ockham replied:
Ah, good catch! That's a leftover from GB times.

I should probably review the PHPDoc at some point 😬

ockham replied:
https://github.com/WordPress/wordpress-develop/pull/5240/commits/d1cf61ed6cbff74eaee4766934ade067d9177d53

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,108 @@ function get_hooked_blocks( $name ) {
+		}
+		array_splice( $parent_block['innerContent'], $chunk_index, 0, array( null ) );
+		return $parent_block;
+	} elseif ( 'last_child' === $relative_position ) {
```
No need to use else anymore because there is the `return` statement:
```suggestion
	}
	if ( 'last_child' === $relative_position ) {
```

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,108 @@ function get_hooked_blocks( $name ) {
 	return $hooked_blocks;
 }
 
+function insert_hooked_blocks( $block ) {
```
Can we land `insert_hooked_block` and tackle `insert_hooked_blocks` in a follow-up? Well, I should start by saying that I don't fully understand the difference between both 😄

ockham replied:
> Can we land `insert_hooked_block` and tackle `insert_hooked_blocks` in a follow-up?

Yes! I was actually thinking the exact same thing! 😄 

> Well, I should start by saying that I don't fully understand the difference between both 😄 

Hehe, I can't blame you, there's a lot of moving parts. I promise it'll all make sense soon! 😊

ockham replied:
940c24bb14

ockham replied:
To explain a bit right away:

`insert_hooked_block` is fairly low-level, and accepts a hooked block for insertion, plus the insertion point specified by the anchor block and relative position (and possibly parent block and block/chunk indices, [see](https://github.com/WordPress/wordpress-develop/pull/5240#issuecomment-1724961426)).

`insert_hooked_blocks` is more high-level: Given a block (found in a template or pattern), it will look up all blocks hooked into it (via `get_hooked_blocks`) and then call `insert_hooked_block` to insert them in the correct locations.

HTH!

gziolo reviewed (commented):
`insert_hooked_block` seems ready to go.

ockham commented:
I was considering doing one more thing: Split `insert_hooked_block` into `insert_hooked_child_block` and `insert_hooked_sibling_block`, since they're pretty different. I also might adjust the function signature of the latter a bit more to align with [this](https://github.com/WordPress/wordpress-develop/pull/5242#discussion_r1329123540).

ockham commented:
> I was considering doing one more thing: Split `insert_hooked_block` into `insert_hooked_child_block` and `insert_hooked_sibling_block`, since they're pretty different. I also might adjust the function signature of the latter a bit more to align with [this](https://github.com/WordPress/wordpress-develop/pull/5242#discussion_r1329123540).

Might actually rename to `insert_child_block` and `insert_sibling_block`, since they're arguably not even that specific to hooked blocks.

ockham commented:
@gziolo Should be ready for review! I've changed it a lot, but I'm really happy with it now. It's much less specific to hooked blocks now, yet the functions are exactly the way we'll need them 😄

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,87 @@ function get_hooked_blocks( $name ) {
+	// when rendering blocks, we also need to insert a value (`null`, to mark a block
+	// location) into that array.
+	array_splice( $parent_block['innerContent'], $chunk_index, 0, array( null ) );
+	return $parent_block;
```
If the preferred approach is to mutate the parent block, then it doesn't need to return the mutated block. `array_splice` is a good example of how it is often handled in PHP.

The same note applies to other helper functions.

ockham replied:
Yeah, that's a good point 🤔 

TBH I don't love mutating functions, but it's kind of a side effect of `array_splice` mutating `$parent_block['innerContent']`; the alternative would be to clone that, and it didn't seem worth the hassle.

I had a feeling that functions that mutate (but don't return) aren't too wide-spread in WP, but maybe I'm wrong. I'll consider removing the return value 👍

ockham replied:
Addressed in 8fee2c6ad3.

gziolo reviewed (approved):
I had a minor note about the return values that seem obsolete for the helper functions because they all mutate the first param, which is then returned. Anyway, I don't have strong feelings about it so if there is any reason to keep them, feel free to leave it as is.

gziolo commented:
Test failures seem unrelated, so maybe there is a need to rebase the PR.

I like how the code is split now into smaller helper functions that are specialized and nicely encapsulate their task. One final note, they will be exposed in the dev docs portal, which I'm fine with.

ockham commented:
> Test failures seem unrelated, so maybe there is a need to rebase the PR.

Yeah, looks like unit tests have been failing for all recent PRs 😕  

> I like how the code is split now into smaller helper functions that are specialized and nicely encapsulate their task. One final note, they will be exposed in the dev docs portal, which I'm fine with.

Yeah, I think that's perfectly okay -- the functions are now generic enough that they might even come in handy for different purposes.

ockham committed:
Blocks: Implement block insertion functions.

For #59313, we need to implement functions to insert a given parsed block into another parsed block's inner blocks, and to prepend and append to that array, respectively.

We will use those functions in combination with `traverse_and_serialize_blocks` (see #59327) to implement automatic insertion of hooked blocks into block templates and patterns.

Props gziolo.
Fixes #59385.

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/56618.
